src/mk.rs

```diff
@@ -0,0 +1,143 @@
+use crate::tree::*;
+use crate::utils::crypto::*;
+use crate::utils::num;
+use std::rc::Rc;
+
+type Hash = [u8; 64];
+type MKNode = TreeNode<Hash>;
+pub struct MerkleTree {
+    tree: MKNode,
+    leaves: Vec<MKNode>,
+    root_hash: Hash,
+}
+
+impl MerkleTree {
+    fn create<T: DataToHash>(data: Vec<T>) -> Option<Self> {
+        if data.is_empty() {
+            return None;
+        }
+        let leaves = MerkleTree::get_leaves_from(data);
+
+        let tree = MerkleTree::create_tree(leaves.clone());
+        let root_hash = tree.borrow().value;
+
+        Some(Self {
+            tree,
+            root_hash,
+            leaves,
+        })
+    }
+
+    fn get_leaves_from<T: DataToHash>(data: Vec<T>) -> Vec<MKNode> {
+        data.iter()
+            .map(|el| Node::new(get_hash_from_data(el), None, None, None))
+            .collect()
+    }
+
+    fn create_tree(mut leaves: Vec<MKNode>) -> MKNode {
+        while leaves.len() != 1 {
+            leaves = leaves
+                .chunks(2)
+                .map(|el| match el {
+                    [a, b] => {
+                        let node = Node::new(
+                            get_combined_hash(a.borrow().value, b.borrow().value),
+                            Some(vec![Rc::clone(a), Rc::clone(b)]),
+                            None,
+                            None,
+                        );
+                        Node::set_parent_and_siblings(&node, &[a, b]);
+                        return node;
+                    }
+                    // hash with itself
+                    [a] => {
+                        let node = Node::new(
+                            get_combined_hash(a.borrow().value, a.borrow().value),
+                            Some(vec![Rc::clone(a)]),
+                            None,
+                            None,
+                        );
+                        // does not have siblings
+                        Node::set_parent(&node, a);
+                        return node;
```

Comment:
You could clone it and set it as sibling of itself, shouldn't impact performance and allows you to always be able to hash an element with it's sibling or call get_siblings without checking if there is Some(node) or None as sibling.

I don't think this is that important tho. And for the root  node you would still have to set sibling to None